sql/core/src/test/scala/org/apache/spark/sql/execution/metric/SQLMetricsSuite.scala

```diff
@@ -504,4 +504,38 @@ class SQLMetricsSuite extends SparkFunSuite with SQLMetricsTestUtils with Shared
   test("writing data out metrics with dynamic partition: parquet") {
     testMetricsDynamicPartition("parquet", "parquet", "t1")
   }
+
+  test("writing metrics from single thread") {
+    val nAdds = 10
+    val acc = new SQLMetric("test", -10)
+    assert(acc.isZero())
+    acc.set(0)
+    for (i <- 1 to nAdds) acc.add(1)
+    assert(!acc.isZero())
+    assert(nAdds === acc.value)
+    acc.reset()
+    assert(acc.isZero())
+  }
+
+  test("writing metrics from multiple threads") {
```

Comment:
Really dumb question, does this fail without the fix?

Reply:
Good question. Checking in a separate branch:
org.scalatest.exceptions.TestFailedException: 100000 did not equal 56544

Reply:
Thanks!

Reply:
FYI, I think this is not a valid test. Spark assumes there should be only one writer at the same time.

Reply:
To be clear, we just need AccumulatorV2 can be read in the heart beat thread. That's the only place need to think about concurrence.

Reply:
Do you mean it's a one-writer, multi-reader scene?

Reply:
This was originally raised because an implementation of broadcast join did have multiple writers.  Unfortunately we recently determined that the LongAdder is causing a performance regression and we are going to revert this.
@cloud-fan or @hvanhovell can one of you send the rollback PR?

Reply:
> Do you mean it's a one-writer, multi-reader scene?

Yes.